numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -3098,11 +3097,11 @@ VOID_compare(char *ip1, char *ip2, PyArrayObject *ap)
             }
             if (swap || !npy_is_aligned(nip2, new->alignment)) {
                 /* create buffer and copy */
-                nip2 = PyDataMem_UserNEW(new->elsize, PyArray_HANDLER(ap)->allocator);
+                nip2 = PyDataMem_UserNEW(new->elsize, current_handler->allocator);
                 if (nip2 == NULL) {
                     if (nip1 != ip1 + offset) {
                         PyDataMem_UserFREE(nip1, new->elsize,
-                                           PyArray_HANDLER(ap)->allocator);
+                                           current_handler->allocator);
```

Comment:
This should still use the (potentially) older handler to match its allocation, no?

Reply:
This is a case of an internal array allocation. Both `nip1` and `nip2` are temporarily allocated inside the function scope.

This means that if we use the current allocator for `UserNEW`, we have to use it for `UserFREE` also. What is currently missing is to ensure that the current allocator won't change in between. But this will be enforced by the context/thread-local feature of the allocator (that will come next).

I think we could add a comment there, explaining.

Reply:
+1 for a comment

Reply:
done.

Feel free to make any changes!